internal/generate/crd/crd.go

```diff
@@ -0,0 +1,348 @@
+	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// crdGenerator configures the CustomResourceDefintion manifest generator
+// for Go and non-Go projects.
+type crdGenerator struct {
+	// Dir containing relevant input files.
+	inputDir string
+	// Dir in which to generate output files.
+	outputDir string
+	// isOperatorGo is true when the operator is written in Go.
+	isOperatorGo bool
+	// resource contains API information used to configure single-CRD generation.
+	// This is only required when isOperatorGo is false.
+	resource scaffold.Resource
+}
+
+// NewGo returns a CRD generator configured to generate CustomResourceDefintion
+// manifests from Go API files.
+func NewGo(inputDir, outputDir string) genutil.Generator {
+	g := crdGenerator{
+		inputDir:     inputDir,
+		outputDir:    outputDir,
+		isOperatorGo: true,
+	}
+	if g.inputDir == "" {
+		g.inputDir = scaffold.ApisDir
+	}
+	if g.outputDir == "" {
+		g.outputDir = scaffold.CRDsDir
+	}
+	g.inputDir = fileutil.DotPath(g.inputDir)
+	g.outputDir = fileutil.DotPath(g.outputDir)
+	return g
+}
+
+// NewGo returns a CRD generator configured to generate a
+// CustomResourceDefintion manifest from scratch using data in resource.
+func NewNonGo(inputDir, outputDir string, resource scaffold.Resource) genutil.Generator {
+	g := crdGenerator{
+		inputDir:     inputDir,
+		outputDir:    outputDir,
+		resource:     resource,
+		isOperatorGo: false,
+	}
+	if g.inputDir == "" {
+		g.inputDir = scaffold.CRDsDir
+	}
+	if g.outputDir == "" {
+		g.outputDir = scaffold.CRDsDir
+	}
+	g.inputDir = fileutil.DotPath(g.inputDir)
+	g.outputDir = fileutil.DotPath(g.outputDir)
+	return g
+}
+
+func (g crdGenerator) validate() error {
+	if g.inputDir == "" {
+		return errors.New("input dir cannot be empty")
+	}
+	if g.inputDir == "" {
+		return errors.New("output dir cannot be empty")
+	}
+	if !g.isOperatorGo && g.resource == (scaffold.Resource{}) {
+		return errors.New("resource cannot be empty")
+	}
+	return nil
+}
+
+// Generate generates CRD manifests and writes them to g.outputDir.
+func (g crdGenerator) Generate() (err error) {
+	if err = g.validate(); err != nil {
+		return errors.Wrap(err, "validation error")
+	}
+	var fileMap map[string][]byte
+	if g.isOperatorGo {
+		fileMap, err = g.generateGo()
+	} else {
+		fileMap, err = g.generateNonGo()
+	}
+	if err != nil {
+		return errors.Wrap(err, "error generating CRD manifests")
+	}
+	if len(fileMap) == 0 {
+		return errors.New("error generating CRD manifests: no generated files found")
+	}
+	if err = os.MkdirAll(g.outputDir, fileutil.DefaultDirFileMode); err != nil {
+		return errors.Wrapf(err, "error mkdir %s", g.outputDir)
+	}
+	for fileName, b := range fileMap {
+		path := filepath.Join(g.outputDir, fileName)
+		if err := ioutil.WriteFile(path, b, fileutil.DefaultFileMode); err != nil {
+			return errors.Wrap(err, "error writing CRD manifests")
+		}
+	}
```

Comment:
wdyt it be writeCRD()?

Reply:
IMO there's not enough code in `Generate()` to warrant refactoring the writing bit into its own function.